src/server/service/debug.rs

```diff
@@ -0,0 +1,220 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error;
+use std::boxed::FnBox;
+use std::fmt::{self, Display};
+
+use grpc::{RpcContext, RpcStatus, RpcStatusCode, ServerStreamingSink, UnarySink};
+use futures::Future;
+use futures::sync::oneshot;
+use kvproto::debugpb_grpc;
+use kvproto::debugpb::*;
+
+use util::worker::{Runnable, Scheduler};
+use raftstore::store::Engines;
+use raftstore::store::debug::*;
+
+use super::make_callback;
+
+pub type Callback = Box<FnBox(Result<Response, Error>) + Send>;
+
+#[derive(Clone)]
+pub struct Service {
+    scheduler: Scheduler<Request>,
+}
+
+impl Service {
+    pub fn new(scheduler: Scheduler<Request>) -> Service {
+        Service { scheduler }
+    }
+
+    fn handle_response<F, M, Q>(
+        &self,
+        ctx: RpcContext,
+        sink: UnarySink<Q>,
+        req: Request,
+        resp: F,
+        map: M,
+        tag: &'static str,
+    ) where
+        Q: 'static,
+        M: FnOnce(Response) -> Q + Send + 'static,
+        F: Future<Item = Result<Response, Error>, Error = oneshot::Canceled> + Send + 'static,
+    {
+        let on_error = move |e| {
+            error!("{} failed: {:?}", tag, e);
+        };
+        if self.scheduler.schedule(req).is_ok() {
+            let future = resp.then(|v| match v {
+                Ok(Ok(resp)) => sink.success(map(resp)).map_err(on_error),
+                Ok(Err(Error::NotFound(msg))) => {
+                    let status = RpcStatus::new(RpcStatusCode::NotFound, Some(msg));
+                    sink.fail(status).map_err(on_error)
+                }
+                Ok(Err(Error::InvalidArgument(msg))) => {
+                    let status = RpcStatus::new(RpcStatusCode::InvalidArgument, Some(msg));
+                    sink.fail(status).map_err(on_error)
+                }
+                Ok(Err(Error::Other(e))) => {
+                    let status = RpcStatus::new(RpcStatusCode::Unknown, Some(format!("{:?}", e)));
+                    sink.fail(status).map_err(on_error)
+                }
+                Err(canceled) => {
+                    let status =
+                        RpcStatus::new(RpcStatusCode::Unknown, Some(format!("{:?}", canceled)));
+                    sink.fail(status).map_err(on_error)
+                }
+            });
+            ctx.spawn(future);
+        } else {
+            let status = RpcStatus::new(RpcStatusCode::Unavailable, None);
+            ctx.spawn(sink.fail(status).map_err(on_error));
+        }
+    }
+}
+
+impl debugpb_grpc::Debug for Service {
+    fn get(&self, ctx: RpcContext, mut req: GetRequest, sink: UnarySink<GetResponse>) {
+        const LABEL: &'static str = "debug_get";
+
+        let (cb, future) = make_callback();
+        let req = Request::Get {
```

Comment:
I think you can use future_cpupool directly, so you don't have to define a lot of `Request`s again.